### PR TITLE
[Core] Support structured outputs for beam search

### PR DESCRIPTION
**Pull request by guan404ming**
## Purpose

Support structured outputs (e.g., JSON schema) for beam search. Previously, beam search had no way to constrain generation to valid JSON or other grammar-defined formats. This adds grammar-based constrained decoding to the beam search loop. 

Related: https://github.com/vllm-project/vllm/issues/34782

## Test Plan

pytest tests/samplers/test_beam_search.py::test_beam_search_structured_output -v

## Test Result

Local test passed

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: vllm/entrypoints/llm.py [hunk 1]
```diff
@@ -104,6 +112,17 @@
...
 
+def _bitmask_to_token_ids(
+    bitmask_row: torch.Tensor, vocab_size: int
+) -> list[int]:
+    """Convert a packed int32 bitmask row to a list of allowed token IDs."""
+    indices = torch.arange(vocab_size)
+    word_indices = indices >> 5  # i // 32
+    bit_indices = indices & 31  # i % 32
+    mask = ((bitmask_row[word_indices] >> bit_indices) & 1).bool()
+    return indices[mask].tolist()
```
[the remark this answers is not in the document]
Comment (guan404ming) [reply vllm/entrypoints/llm.py]: Cached indices, word_indices, bit_indices by vocab_size

File: vllm/entrypoints/llm.py [hunk 2]
```diff
@@ -825,6 +949,125 @@ def beam_search(
...
+                vocab_size=vocab_size,
+            )
+        else:
+            raise ValueError(
+                f"Unsupported structured output backend: {backend_name}"
+            )
+
+        structured_output_key = get_structured_output_key(structured_outputs)
+        request_type, grammar_spec = structured_output_key
+        backend.compile_grammar(request_type, grammar_spec)
```
[the remark this answers is not in the document]
Comment (guan404ming) [reply vllm/entrypoints/llm.py]: Removed it. _build_beam_sampling_params compiles per-beam anyway

File: vllm/entrypoints/llm.py [hunk 3]
```diff
@@ -825,6 +949,125 @@ def beam_search(
...
+        request_type, grammar_spec = structured_output_key
+        result: list[SamplingParams | None] = []
+
+        for beam in beams:
+            grammar = backend.compile_grammar(request_type, grammar_spec)
+            prompt_len = len(beam.orig_prompt["prompt_token_ids"])
+            generated_tokens = beam.tokens[prompt_len:]
+
+            if generated_tokens:
+                grammar.accept_tokens("beam", generated_tokens)
```
[the remark this answers is not in the document]
Comment (guan404ming) [reply vllm/entrypoints/llm.py]: Backends don't support cloning grammar state. Added comment explaining why replay is needed

File: vllm/entrypoints/llm.py [hunk 4]
```diff
@@ -738,77 +769,170 @@ def beam_search(
...
+
+                    allowed_sets: list[set[int] | None] = [
+                        None
+                    ] * len(all_beams)
+                    if structured_output_backend is not None:
+                        for idx, p in zip(active_indices, active_params):
+                            if p is not None and p.allowed_token_ids:
+                                allowed_sets[idx] = set(
+                                    p.allowed_token_ids
+                                )
```
[the remark this answers is not in the document]
Comment (guan404ming) [reply vllm/entrypoints/llm.py]: The logprob filtering is necessary. In v1/sample/sampler.py, raw_logprobs is captured at line 82 (before any masking), while allowed_token_ids_mask is applied at line 287 to logits for sampling only. The top-k logprobs returned to the caller include tokens outside allowed_token_ids. Without this filter, grammar.accept_tokens() fails on disallowed tokens. Verified empirically during testing.

File: vllm/entrypoints/llm.py [hunk 5]
```diff
@@ -738,77 +775,152 @@ def beam_search(
...
-                    "reflect instance-level progress."
-                )
-            for _ in token_iter:
-                all_beams: list[BeamSearchSequence] = list(
-                    sum((instance.beams for instance in instances_batch), [])
-                )
-                pos = [0] + list(
-                    itertools.accumulate(
-                        len(instance.beams) for instance in instances_batch
+        try:
```
Comment (DarkLight1337) [inline vllm/entrypoints/llm.py]: I think this method is getting a bit long, could we factor this out?

Comment (guan404ming) [reply vllm/entrypoints/llm.py]: Thanks for the review. 

Sure, that makes sense to me. Just extracted `_beam_search_step` method from the inner loop of `beam_search`. The `beam_search` method now just calls `self._beam_search_step(...)` inside the `for _ in token_iter` loop

File: vllm/entrypoints/llm.py [hunk 6]
```diff
@@ -801,6 +828,254 @@ def beam_search(
...
+        structured_outputs: StructuredOutputsParams,
+        tokenizer: TokenizerLike,
+    ) -> tuple[StructuredOutputBackend, tuple, torch.Tensor]:
+        """Initialize the structured output backend for beam search."""
+        temp_params = SamplingParams(
+            max_tokens=1,
+            structured_outputs=structured_outputs,
+        )
+        vllm_config = self.llm_engine.vllm_config
+        temp_params.verify(
```
Comment (DarkLight1337) [inline vllm/entrypoints/llm.py]: Why do we need to verify this here? Shouldn't this be handled in `InputPreprocessor` when the request is added anyway?

File: vllm/entrypoints/llm.py [hunk 7]
```diff
@@ -801,6 +828,254 @@ def beam_search(
...
+            model_config=vllm_config.model_config,
+            speculative_config=vllm_config.speculative_config,
+            structured_outputs_config=vllm_config.structured_outputs_config,
+            tokenizer=tokenizer,
+        )
+
+        backend_name = structured_outputs._backend
+        vocab_size = self.model_config.get_vocab_size()
+
+        backend: StructuredOutputBackend
```
Comment (DarkLight1337) [inline vllm/entrypoints/llm.py]: I think this is repeating too much from `StructuredOutputManager`. How should we handle this? @njhill @russellb

File: vllm/entrypoints/llm.py (hunk 6 above)
Comment (guan404ming) [reply vllm/entrypoints/llm.py]: I originally called verify() to resolve the backend name from the engine's `structured_outputs_config`, but you're right that it also runs a bunch of unrelated validations (logprobs, logit_bias, etc.) and the actual request-level validation is already handled by `InputPreprocessor`. Replaced it with a direct `_backend` assignment from `structured_outputs_config`.

File: vllm/entrypoints/llm.py (hunk 7 above)
Comment (guan404ming) [reply vllm/entrypoints/llm.py]: Should we extract a shared factory function here?

**Comment by guan404ming** [comment]
@DarkLight1337 Friendly ping. All comments have been addressed and the PR is rebased on the latest main with conflicts resolved. PTAL when you get a chance. cc @russellb in case you have bandwidth to take a look at the structured outputs side. Thanks!

**Comment by guan404ming** [comment]
Gentle ping @DarkLight1337, rebased onto latest main and resolved conflicts. PTAL, thanks!

**Review by DarkLight1337 — APPROVED** [review]
Now that this has been factored out into a mixin I'm more comfortable with introducing this, thanks and sorry for the delay!

**Comment by guan404ming** [comment]
No worries, thanks for the review!

**Comment by DarkLight1337** [comment]
Please fix the failing test: https://buildkite.com/vllm/ci/builds/71723/canvas?jid=019ebb53-25c4-412d-bd2b-e22b141b128c&tab=output

**Comment by guan404ming** [comment]
Thanks @DarkLight1337!
